dwave/plugins/torch/samplers/block_spin_sampler.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2025 D-Wave
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from typing import TYPE_CHECKING, Callable, Literal
+
+import torch
+from torch import nn
+
+from dwave.plugins.torch.nn.functional import bit2spin_soft
+from dwave.plugins.torch.models.boltzmann_machine import GraphRestrictedBoltzmannMachine as GRBM
+from dwave.plugins.torch.tensor import rands
+
+__all__ = ["BlockSpinSampler"]
+
+
+class BlockSpinSampler(nn.Module):
+    """A block-spin update sampler for graph-restricted Boltzmann machines.
+
+    Note this is a sampler and not a neural network. It extends `nn.Module` for the convenience of
+    managing devices of indices (stored as parameters).
+
+    Due to the sparse definition of GRBMs, some tedious, and ugly, indexing tricks are required for
+    efficiently sampling in blocks of spins. Ideally, an adjacency list can be used, however,
+    adjacencies are ragged, making vectorization inapplicable.
+
+    Block-Gibbs and Block-Metropolis obey detailed balance and are ergodic methods at finite
+    temperature which, at fixed parameters, converge upon Boltzmann distributions. Block-Metropolis
+    allows higher acceptance rates for proposals (faster single-step mixing), but is non-ergodic in
+    the limit of zero temperature. Decorrelation from an initial condition can be slower.
+    Block-Gibbs represents best practice for independent sampling.
+
+    Args:
+        grbm (GRBM): The Graph-Restricted Boltzmann Machine to sample from.
+        crayon (Callable): A colouring function; a function that maps a single node of the `grbm` to
+            its colour.
+        num_chains (int): Number of Markov chains to run in parallel.
+        proposal_acceptance_criteria (Literal["Gibbs", "Metropolis"]): The proposal acceptance
+            criterion used to accept or reject states in the Markov chain. Defaults to "Gibbs".
+        seed (Optional[int]): Random seed. Defaults to None.
+    """
+
+    def __init__(self, grbm: GRBM, crayon: Callable, num_chains: int,
+                 proposal_acceptance_criteria: Literal["Gibbs", "Metropolis"] = "Gibbs", seed=None):
+        super().__init__()
+        if num_chains < 1 or not isinstance(num_chains, int):
```

Comment:
python performs these checks in order from left to right. Therefore, if you pass `num_chains="hello world"`, a different error will be thrown saying that you cannot compare order between a str with an int. First check type, then check value.